### ISSUE TITLE
Naming layers

### ISSUE THREAD

**Issue by dimatura** [benanne/Lasagne#115]
This arose in the context of issue #111. One issue I've had with `lasagne` is easily identifying layers and their respective (hyper)parameters. This comes up when configuring layers, but also when serializing them, analyzing them after training, and performing 'surgery' on nets for various purposes, such as discriminative pretraining. 

Currently I rely on creating my own dictionaries with name keys and used that to manage them, but it gets messy sometimes to correlate my own names with what e.g., `get_all_param_values` returns.  I feel like simply having a `name` member for the base layer class would make this somewhat easier. I don't necessarily think  they should have a functional role; they can be mostly for debugging/usabilty, and defaulting to `None`, similar to the ways names work for `theano` variables.

**Comment by benanne** [comment]
Personally I haven't had much use for this, but I think this feature has plenty of use cases that are worth supporting. It's easy enough to add support for it to the base class. Unfortunately we will have to update all existing layer classes to pass the optional 'name' argument to the superclass constructor, which is a bit unfortunate. (Luckily users won't have to update their custom layer classes if they don't want to at least - the names of their layer instances will just be `None`). I guess it should be the second argument on the `Layer` constructor, and preferably used as a keyword argument.

It does add a tiny bit of overhead to creating your own `Layer` subclass, but I suppose it's not excessive (and it can safely be ignored, as long as you're not planning to send a pull request).

Maybe we should consider having all `Layer` classes pass `*args` and `**kwargs` to the base class instead? Just as a way of future-proofing the code. We currently already use this scheme with `get_output()` and `get_output_for()`, to support flags and optional arguments that only affect certain layers in a network (for example, `deterministic=True` to turn off noise generating layers). I can't really think of any other use cases for this at the moment (besides this one), but maybe that's kind of the point :)

Anyway, I'm fine with adding a `name` argument. What does everyone else think? @f0k usually has insightful stuff to say about things like this ;)

**Comment by craffel** [comment]
I have definitely thought this too. I think layers should take an optional name argument, and their parameters should have names based on this name too. E.g., a dense layer with name "blah" would have its weight matrix shared named "blah_W" or something.

**Comment by dimatura** [comment]
Yes, definitely a keyword argument.

@craffel Funny you say that, I use that same convention and I was going to propose it maybe later down the line ;)

Passing on `name` is a bit annoying, and `**kwargs` could also make sense. I guess there's all sorts of metadata people could attach to layers (e.g. for different learning rates or for different regularizations) but since that's pretty application dependent it might be better managed externally.

**Comment by dnouri** [comment]
On 01/25/2015 09:35 PM, Sander Dieleman wrote:
> Maybe we should consider having all |Layer| classes pass |*args| and
> |**kwargs| to the base class instead? Just as a way of future-proofing
> the code.

I agree that this is a good idea and it'll make the code more
future-proof.  **kwargs are much more useful for this than variable
positional *args though.  I'd see if there is really an example where
it's useful to have *args passed to the base class, or maybe get rid of it.

> Anyway, I'm fine with adding a |name| argument. What does everyone else
> think?

I like name because it makes it easier to build an API around layers.  :-)

**Comment by benanne** [comment]
What bothers me about giving layers names like this is that you're effectively giving them two names: a string passed to the constructor, and the name of the variable that contains it. This is basically duplicate code: there are now two "namespaces" that layers live in: the python variable namespace, and the string based namespace we've created. Some parts of the code will refer to layers using the former, others using the latter. That just feels a bit icky, IMO.

Unfortunately I don't think there is a better solution: there is no way to access the name of a variable programmatically, and even if there was it would be rather hackish to use that approach.

I agree that naming parameter variables after their containing layers automatically is a good idea.

**Comment by benanne** [comment]
By the way, the reason that get_output and friends also have *args in addition to **kwargs is that you might want to pass additional inputs that have some sort of separate semantics from the main input for certain layers. It just feels more natural to use positional arguments for that. It's hard to think of an example though!

**Comment by dimatura** [comment]
Yes, the double name issue is a bit bothersome. Seems like the latest `theano` has been working on autonaming their variables from the code, but I think it is rather hackish too. 

In the end the two names play different roles, I think. In code the variable name is more about the actual implementation for training/testing; the other name is more about interfacing with the outside world, for configuration, text display, exporting, etc. which might happen outside of `lasagne`, and even python.

**Comment by craffel** [comment]
> What bothers me about giving layers names like this is that you're effectively giving them two names: a string passed to the constructor, and the name of the variable that contains it. 

Yes, this is true, but it's the same convention/pattern used by theano and I don't know if any non-hacky way to do it otherwise. And there are use cases for these two different names to be different.

**Comment by benanne** [comment]
Sure. I'm definitely not against it, I was just trying to give some background for why this feature doesn't exist yet. I was hoping we could do without it, but I don't think we can.

However, I think we should clearly separate the namespaces, as discussed before. When referring to a layer in the code, we should always use the instances, and not their string names. For example when calling `get_output()` with a dictionary mapping layer instances to Theano expressions, the keys in this dictionary should be actual `Layer` instances, not strings. We should do this consistently throughout the library. As @dimatura said, string names should not have a functional role.

**Comment by craffel** [comment]
> However, I think we should clearly separate the namespaces, as discussed before. 

Yeah, definitely agree here.

**Comment by f0k** [comment]
I'm fine with giving them names, too. As you all already agreed on, I also think they shouldn't have any functional role, just as in Theano. They would just be used for inspection, maybe for error messages or pretty-printing a network architecture.

> I guess it should be the second argument on the Layer constructor, and preferably used as a keyword argument.

Yes. It shouldn't be the second argument on subclass constructors, though, because we still need subclasses to define required positional arguments (such as `num_units` for the `DenseLayer`). Maybe the last keyword argument then, to encourage users to really use the keyword argument form? Passing `kwargs` from subclasses to superclasses would be possible, but we'd need to be careful not to hide spelling errors (someone instantiating a `DropoutLayer` with `P=0.7`, for example). If the base class doesn't get a `**kwargs`, this should work.

**Comment by benanne** [comment]
Okay, so then we would give `Layer.__init__()` an extra keyword argument `name`, and all `Layer` subclasses are modified to have their `__init__()` pass `*args` and `**kwargs` to their superclass. Am I understanding everything correctly? That way if we ever decide to add more arguments to the base class constructor, modifications to the subclasses are no longer necessary, and users don't have to remember to pass every additional argument to the base class manually.

I don't really see a point in passing only `**kwargs`, since the common 'idiom' is to pass both anyway. But maybe there's a good reason to leave out `*args`?

**Comment by dnouri** [comment]
On 01/26/2015 02:00 PM, Sander Dieleman wrote:
> Okay, so then we would give |Layer.__init__()| an extra keyword argument
> |name|, and all |Layer| subclasses are modified to have their
> |__init__()| pass |*args| and |**kwargs| to their superclass. Am I
> understanding everything correctly? That way if we ever decide to add
> more arguments to the base class constructor, modifications to the
> subclasses are no longer necessary, and users don't have to remember to
> pass every additional argument to the base class manually.
> 
> I don't really see a point in passing only |**kwargs|, since the common
> 'idiom' is to pass both anyway. But maybe there's a good reason to leave
> out |*args|?

I'd say the reason is that *args aren't really suited for making
things more robust to future change.

Let's compare the base Layer and DenseLayer:

    class Layer(object):
        def __init__(self, input_layer):
            self.input_layer = input_layer

    class DenseLayer(Layer):
        def __init__(self, input_layer, num_units,
                     W=init.Uniform(), b=init.Constant(0.),
                     nonlinearity=nonlinearities.rectify):
            super(DenseLayer, self).__init__(input_layer)

Now imagine that in some future version, Layer decides to add another
positional argument without a default, say 'foo'.  Now take a look at
a typical instantiation of DenseLayer:

    layer = DenseLayer(input_layer, num_units)

Surely we can add 'foo' to the list and have everything continue to
work without too many changes?

    layer = DenseLayer(input_layer, foo, num_units)

Not this way; this will obviously need a change to the DenseLayer
signature, so that 'foo' comes before 'num_units', so nothing gained.

What about:

    layer = DenseLayer(input_layer, num_units, foo)

Nope, that'll pass foo as 'W' into the DenseLayer's __init__.  Doesn't
work.

I don't think you'll see positional arguments used a lot in the
context of preserving backward compatibility.  From that point of
view, and if noone can come up with a use case, they're just noise
that adds to cognitive overhead.

**Comment by f0k** [comment]
> I don't really see a point in passing only `**kwargs`, since the common 'idiom' is to pass both anyway. But maybe there's a good reason to leave out `*args`?

Hmm, probably not. It would force people to use `name=...` so we'd be free to change the order of extra arguments in the base class at any time, but it's not really necessary, no.

**Comment by benanne** [comment]
That makes a lot of sense. So I think we're in agreement that:
* the `Layer` base class should receive a `name` argument that defaults to `None`. It will simply set the `name` property on the `Layer` class for now.
* All `Layer` subclasses that override `__init__()` (which is most of them) should be modified so that the constuctor takes `**kwargs` and passes these to the base class constructor.

If that's all I can whip up a pull request for this today. Or someone else can do it if they want to, I don't imagine it's a lot of work.

One other thing we could consider is adding `__str__` to the `Layer` base class, using the name if it is available (and maybe also `__repr__` while we're at it). But maybe that's a separate issue.

**Comment by dnouri** [comment]
On 01/28/2015 11:03 AM, Sander Dieleman wrote:

> All Layer classes that have parameters should name those parameters, 
> and if the layer has a name, this should be used as a prefix for the 
> parameter names. This will require modifications to 
> Layer.create_param().

I don't understand the "[layer name] should be used as a prefix for the
parameter names" bit.  (Maybe because I haven't used create_param
before.)  Could you elaborate?

**Comment by benanne** [comment]
The idea is that a `DenseLayer` will name its parameters `W` and `b` by default (to match the variable names), but if the layer has a name, let's say "foo", it should name them `foo_W` and `foo_b` instead.

**Comment by f0k** [comment]
Sounds good to me! (Except that maybe `foo.W` and `foo.b` would be more intuitive? I don't have a strong opinion on that. I can just imagine that both the layer name and the parameter names could contain underscores themselves already, and Theano's variable names do not need to be valid Python variable names anyway.)

**Comment by benanne** [comment]
Yeah, I guess that is more intuitive.

**Comment by benanne** [comment]
I implemented those three things in #117. I modified `create_params` so that it prefixes the parameter names with the layer name automatically. The only case where we can't name the parameter variables is when the user passes in a shared variable, but I think that's okay.

(We could also simply assign to the `name` attribute of the variable, but I think that might be unexpected. If a user names a variable, they probably want to use that specific name and we shouldn't override it.)

This means that `create_param` is no longer a static method.
